### PR TITLE
Strict Null Checks Final

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Finishes implementing Strict Null Checks

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -323,9 +323,10 @@ module.exports = {
       },
     },
     {
-      files: ["**/*.test.ts?(x)"],
+      files: ["**/*.test.ts?(x)", "**/*.spec.ts"],
       rules: {
         "@typescript-eslint/no-unnecessary-type-assertion": "off",
+        "@typescript-eslint/no-non-null-assertion": "off",
```
Comment (grahamlangford) [inline .eslintrc.js:329]: Allow use of `!` in e2e test files

File: src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx [hunk 2]
```diff
@@ -167,11 +169,11 @@ const dataFactory = ({
...
     label: "Zapier - use to connect to PixieBrix from Zapier",
-    integrationId: null,
+    integrationId: null as unknown as RegistryId,
     _rawIntegrationConfigBrand: null,
-    id: null,
-    config: null,
-    integration: null,
+    id: null as unknown as UUID,
+    config: null as unknown as SecretsConfig,
+    integration: null as unknown as Integration,
```
Comment (grahamlangford) [inline src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx:176]: Rather than loosen the type, since this works everywhere currently, I went with casting

**Comment by grahamlangford** [comment]
Should've opened a PR with just enforcing all files in tsconfig.strictNullChecks first. Didn't realize deleting it was going to cause so many other changes to be required. Most file changes were lint auto-fixes

File: .eslintrc.js [hunk 3]
```diff
@@ -323,9 +323,10 @@ module.exports = {
       },
     },
     {
-      files: ["**/*.test.ts?(x)"],
+      files: ["**/*.test.ts?(x)", "**/*.spec.ts"],
```
Comment (fungairino) [inline .eslintrc.js:326]: We also use `**/*.setup.ts` for the setup e2e project files

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -149,8 +149,6 @@ jobs:
           cache: npm
       - run: npm ci
       - run: npm run build:typescript
-      # Prevent regressions in the files that have already been made strictNullCheck-compliant
-      - run: npm run build:strictNullChecks
```
Comment (fungairino) [inline .github/workflows/ci.yml:153]: 🎉

**Comment by twschiller** [comment]
image

File: end-to-end-tests/tests/modLifecycle.spec.ts [hunk 5]
```diff
@@ -111,7 +111,7 @@ test("create, run, package, and update mod", async ({
   });
 
   await test.step("View and update mod in the Workshop", async () => {
-    const workshopPage = new WorkshopPage(newPage, extensionId);
+    const workshopPage = new WorkshopPage(newPage!, extensionId);
```
Comment (fungairino) [inline end-to-end-tests/tests/modLifecycle.spec.ts:114]: Would be nice to avoid the `!` here and in other similar places where we assign variables in a `test.step`. Could we instead assert that the variable is not null right after the step?

File: eslint-local-rules/noRestrictedSyntax.tsx [hunk 6]
```diff
@@ -3,16 +3,15 @@
 // Ok
 import React from "react";
 
-// eslint-disable-next-line no-restricted-imports
+// eslint-disable-next-line no-restricted-imports -- ok for test file
```
Comment (fungairino) [inline eslint-local-rules/noRestrictedSyntax.tsx:6]: nit: I wouldn't call a lint rule a test personally.

**Comment by twschiller** [comment]
@grahamlangford instead of casting thrown error variables from `unknown` to `Error`, I would consider wrapping in a `castError` or similar to convert null/undefined into an error object not already an error

But can be done in a FLUP PR

File: src/auth/RequireAuth.tsx [hunk 7]
```diff
@@ -181,7 +181,7 @@ const RequireAuth: React.FC<RequireAuthProps> = ({
 
   // `useRequiredAuth` handles 401 and other auth-related errors. Rethrow any other errors, e.g., internal server error
   if (meError && !ignoreApiError) {
-    throw meError;
+    throw meError as Error;
```
Comment (fungairino) [inline src/auth/RequireAuth.tsx:184]: You should be able to just thrown an unknown type right? Why the need for `as Error`?

File: src/pageEditor/fields/RemoteMethodOptions.tsx [hunk 8]
```diff
@@ -64,34 +63,40 @@ const RemoteMethodOptions: React.FunctionComponent<
 
   return (
     <div>
-      {/* eslint-disable @typescript-eslint/no-unnecessary-type-assertion -- strictNullChecks requires the cast */}
+      {}
```
Comment (fungairino) [inline src/pageEditor/fields/RemoteMethodOptions.tsx:66]: nit: remove this

File: src/auth/RequireAuth.tsx (hunk 7 above)
Comment (twschiller) [reply src/auth/RequireAuth.tsx:184]: There's a lint rule on currently to force throwing things that are only known to be objects: https://github.com/pixiebrix/pixiebrix-extension/pull/9049#issuecomment-2307842296

File: end-to-end-tests/tests/modLifecycle.spec.ts (hunk 5 above)
Comment (BLoe) [reply end-to-end-tests/tests/modLifecycle.spec.ts:114]: Could also set it to the current page initially to avoid the optional typing?

File: src/pageEditor/layout/ModComponentFormStateWizard.tsx [hunk 9]
```diff
@@ -61,7 +61,7 @@ const WizardNavItem: React.FunctionComponent<{
 const ModComponentFormStateWizard: React.FunctionComponent<{
   modComponentFormState: ModComponentFormState;
 }> = ({ modComponentFormState }) => {
-  const [step, setStep] = useState(wizard[0].step);
```
Comment (fungairino) [inline src/pageEditor/layout/ModComponentFormStateWizard.tsx:64]: would recommend to `as const` `wizard` instead.

File: src/pageEditor/tabs/editTab/EditTab.tsx [hunk 10]
```diff
@@ -90,6 +91,10 @@ const EditTab: React.FC<{
               name="copyNode"
               icon={faCopy}
               onClick={() => {
+                assertNotNullish(
+                  activeNodeId,
+                  "activeNodeId is required to copy a node",
+                );
```
Comment (fungairino) [inline src/pageEditor/tabs/editTab/EditTab.tsx:97]: what happens if a click handler fails? Does the error boundary still catch this?

Comment (fungairino) [reply src/pageEditor/tabs/editTab/EditTab.tsx:97]: Yea, in this case it wouldn't be caught by the errorBoundary. In this case you might want to do an `if condition` instead

File: src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx [hunk 11]
```diff
@@ -351,20 +355,20 @@ const usePipelineNodes = (): {
...
         }
       : undefined;
 
-    const hovered = hoveredState[blockConfig.instanceId];
+    const hovered = hoveredState[instanceId];
     const onHoverChange = (hovered: boolean) => {
       setHoveredState((previousState) => ({
         ...previousState,
-        [blockConfig.instanceId]: hovered,
+        [instanceId as string]: hovered,
```
Comment (fungairino) [inline src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx:371]: I guess you need to cast a tagged string type to string for object key assignments?

File: src/types/helpers.ts [hunk 12]
```diff
@@ -71,9 +75,6 @@ export const UNSET_UUID = validateUUID("00000000-0000-4000-A000-000000000000");
 
 export function validateUUID(uuid: unknown): UUID {
   if (uuid == null) {
-    /* eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion --
-    We don't have strictNullChecks on, so null values will find there way here. We should pass them along. Eventually
-    we can remove this check as strictNullChecks will check the call site */
     return uuid as unknown as UUID;
```
Comment (fungairino) [inline src/types/helpers.ts:78]: per the old comment, do we still need this function, then?

File: tsconfig.json [hunk 13]
```diff
@@ -11,7 +11,6 @@
     "outDir": null,
 
     // TODO: Drop these lines to make TS stricter https://github.com/pixiebrix/pixiebrix-extension/issues/775
-    "strictNullChecks": false,
```
Comment (fungairino) [inline tsconfig.json:14]: 🎉

File: src/pageEditor/fields/RemoteMethodOptions.tsx (hunk 8 above)
Comment (BLoe) [inline src/pageEditor/fields/RemoteMethodOptions.tsx:66]: Looks like this was an automatic change? The braces got left in here

File: src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx [hunk 14]
```diff
@@ -215,13 +217,14 @@ const PrivateIntegrationsCard: React.FunctionComponent<OwnProps> = ({
...
-    [forceShowIntegrationConfigId],
-  );
+  const forceShowRecord: ((config: IntegrationConfig) => boolean) | undefined =
+    useMemo(
+      () =>
+        forceShowIntegrationConfigId
+          ? ({ id }) => id === forceShowIntegrationConfigId
+          : undefined,
+      [forceShowIntegrationConfigId],
+    );
```
Comment (BLoe) [inline src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx:227]: This use-case of the prop being a predicate and not just a row id is not actually being handled properly in the child component: https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/components/paginatedTable/PaginatedTable.tsx#L209

File: src/types/helpers.ts [hunk 15]
```diff
@@ -60,7 +60,11 @@ export function uuidv4(): UUID {
   return v4() as UUID;
 }
 
-export function isUUID(uuid: string): uuid is UUID {
+export function isUUID(uuid: string | undefined): uuid is UUID {
```
Comment (BLoe) [inline src/types/helpers.ts:63]: Type guard is generally `unknown` for the input param type, right?

File: src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx [hunk 16]
```diff
@@ -426,7 +431,8 @@ const usePipelineNodes = (): {
         }),
         brickLabel: isNullOrBlank(blockConfig.label)
           ? block?.name
-          : blockConfig.label,
+          : // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- checked by isNullOrBlank
+            blockConfig.label!,
```
Comment (BLoe) [inline src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx:435]: Do we need the bang at all here? Isn't the other ternary branch returning a possibly-undefined value as well?

File: src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx [hunk 17]
```diff
@@ -553,12 +559,14 @@ const usePipelineNodes = (): {
...
-            ? isHeaderNodeActive
-            : isParentActive || isAncestorActive,
+          isParentActive:
+            (isSiblingHeaderActive
+              ? isHeaderNodeActive
+              : isNodeActive || isParentActive) || false,
+          isAncestorActive:
+            (isSiblingHeaderActive
+              ? isHeaderNodeActive
+              : isParentActive || isAncestorActive) || false,
```
Comment (BLoe) [inline src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx:569]: Nit: This nested logic is pretty hard to parse at a glance when reading the code

File: src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx [hunk 18]
```diff
@@ -576,6 +584,11 @@ const usePipelineNodes = (): {
         modComponentHasTraces ||= subPipelineHasTraces;
       }
 
+      assertNotNullish(
+        blockConfig.outputKey,
+        "outputKey is required for footer",
+      );
```
Comment (BLoe) [inline src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx:590]: Are we sure this is true?

File: .eslintrc.js (hunk 3 above)
Comment (grahamlangford) [reply .eslintrc.js:326]: I'm not sure that's necessary. We'll probably have project specific eslint files for the mono-repo, I would delay until then. This is just turning off some rules anyway

File: eslint-local-rules/noRestrictedSyntax.tsx (hunk 6 above)
Comment (grahamlangford) [reply eslint-local-rules/noRestrictedSyntax.tsx:6]: Oops you're right. I was thinking this was the file we use to _test_ noRestrictedSyntax. I'll update the comment

File: end-to-end-tests/tests/modLifecycle.spec.ts (hunk 5 above)
Comment (grahamlangford) [reply end-to-end-tests/tests/modLifecycle.spec.ts:114]: Tried with assertNotNullish, but it didn't work. I'm not sure why

**Comment by grahamlangford** [comment]
> @grahamlangford instead of casting thrown error variables from unknown to Error, I would consider wrapping in a castError or similar to convert null/undefined into an error object not already an error to match the spirit of the rule

> But could be done in a future PR

> Some discussion on their issue tracker here: https://github.com/typescript-eslint/typescript-eslint/issues/4206#issuecomment-976632348

Created follow-on ticket here: https://github.com/pixiebrix/pixiebrix-extension/issues/9064

File: src/auth/RequireAuth.tsx (hunk 7 above)
Comment (grahamlangford) [reply src/auth/RequireAuth.tsx:184]: It's a lint rule we need to handle better: https://github.com/pixiebrix/pixiebrix-extension/issues/9064

File: src/pageEditor/layout/ModComponentFormStateWizard.tsx (hunk 9 above)
Comment (grahamlangford) [reply src/pageEditor/layout/ModComponentFormStateWizard.tsx:64]: Still have to include a `!`, but slightly better

File: src/pageEditor/tabs/editTab/EditTab.tsx (hunk 10 above)
Comment (grahamlangford) [reply src/pageEditor/tabs/editTab/EditTab.tsx:97]: It's a scenario that realistically should never happen. There wouldn't be a copy brick button if there's not an active node. If that did somehow happen it _should_ error

File: src/types/helpers.ts (hunk 12 above)
Comment (grahamlangford) [reply src/types/helpers.ts:78]: We haven't handled upstream properly. We should probably look to eliminate the need for this function in a future DevEx ticket

File: src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx (hunk 14 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx:227]: Are you sure? Looks right to me?

File: src/types/helpers.ts (hunk 15 above)
Comment (grahamlangford) [reply src/types/helpers.ts:63]: Only if it's _really_ unknown. It's very typical for something like this where we want to typeguard on the type of string

File: src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx (hunk 17 above)
Comment (grahamlangford) [reply src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx:569]: Totally agree, but this file is due to be heavily refactored soon, so I didn't want to spend time on it

File: src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx (hunk 18 above)
Comment (grahamlangford) [reply src/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes.tsx:590]: I thought I did confirm that. But no, it's not. Fixed

File: src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx (hunk 14 above)
Comment (BLoe) [reply src/extensionConsole/pages/integrations/PrivateIntegrationsCard.tsx:227]: Huh weird, I guess it is right, not sure what I was looking at, I think I thought `findPageIndex` was a library function or something 🤷‍♂️

**Review by BLoe — APPROVED** [review]
